src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -7,13 +7,220 @@
+                                   UCS_SOCKADDR_STRING_LEN));
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ucs_debug("rdmacm_cm ep %p (id=%p) disconnecting from peer :%s", cep, cep->id,
+              ucs_sockaddr_str(rdma_get_peer_addr(cep->id), ip_port_str,
+                               UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+uct_base_iface_t dummy_iface = {
+    .super = {
+        .ops = {
+            .ep_pending_purge = (void *)ucs_empty_function_return_success,
+            .ep_disconnect    = uct_rdmacm_cm_ep_disconnect,
+            .ep_destroy       = UCS_CLASS_DELETE_FUNC_NAME(uct_rdmacm_cm_ep_t)
+        }
+    }
+};
+
+UCS_CLASS_INIT_FUNC(uct_rdmacm_cm_ep_t, const uct_ep_params_t *params)
+{
+    ucs_status_t status;
+
+    if (!(params->field_mask & UCT_EP_PARAM_FIELD_CM)) {
+        ucs_print("No UCT_EP_PARAM_FIELD_CM set: %zu", params->field_mask);
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    if (!(params->field_mask & UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS) ||
+        !(params->sockaddr_cb_flags & UCT_CB_FLAG_ASYNC)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (!(params->field_mask & (UCT_EP_PARAM_FIELD_SOCKADDR |
+                                UCT_EP_PARAM_FIELD_CONN_REQUEST))) {
+        ucs_print("Neithter UCT_EP_PARAM_FIELD_SOCKADDR or UCT_EP_PARAM_FIELD_CONN_REQUEST set: %zu",
+                 params->field_mask);
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &dummy_iface);
+
+    self->cm                  = ucs_derived_of(params->cm, uct_rdmacm_cm_t);
+    self->wireup.priv_pack_cb = (params->field_mask &
+                                 UCT_EP_PARAM_FIELD_SOCKADDR_PACK_CB) ?
+                                params->sockaddr_pack_cb : NULL;
+    self->disconnect_cb       = (params->field_mask &
+                                 UCT_EP_PARAM_FIELD_SOCKADDR_DISCONNECT_CB) ?
+                                params->disconnect_cb : NULL;
+    self->user_data           = (params->field_mask &
+                                 UCT_EP_PARAM_FIELD_USER_DATA) ?
+                                params->user_data : NULL;
+    self->cq                  = NULL;
+    self->qp                  = NULL;
+
+    if (params->field_mask & UCT_EP_PARAM_FIELD_SOCKADDR) {
+        status = uct_rdamcm_cm_ep_client_init(self, params);
+    } else {
+        ucs_assert(params->field_mask & UCT_EP_PARAM_FIELD_CONN_REQUEST);
+        status = uct_rdamcm_cm_ep_server_init(self, params);
+    }
+
+    if (status == UCS_OK) {
+        ucs_debug("created an RDMACM endpoint %p on CM %p. rdmacm_id: %p",
+                  self, self->cm, self->id);
+    }
+
+    return status;
+}
+
 UCS_CLASS_CLEANUP_FUNC(uct_rdmacm_cm_ep_t)
 {
+    uct_priv_worker_t *worker_priv;
+    int ret;
+
+    worker_priv = ucs_derived_of(self->cm->worker, uct_priv_worker_t);
+
+    UCS_ASYNC_BLOCK(worker_priv->async);
+
+    if (self->qp != NULL) {
```

Comment:
aren't qp and cq always created?

Reply:
in a 'good' flow yes, but if ep_destroy is called before the event that triggers their creation (route_resolved on the client side), it will segfault without this check.